python/tank/path_cache.py

```diff
@@ -1265,3 +1170,142 @@ def get_secondary_entities(self, path):
+        
+        # free up memory
+        sg_data = None
+                
+        
+        cursor = self._connection.cursor()
+
+        try:
+            # get all records and check each one against shotgun.
+            pc_data = list(cursor.execute("""select pc.rowid,
+                                                    pc.entity_type, 
+                                                    pc.entity_id, 
+                                                    pc.entity_name, 
+                                                    pc.root, 
+                                                    pc.path, 
+                                                    pc.primary_entity 
+                                             from path_cache pc"""))
+        finally:
+            cursor.close()
+        
+        log.info("")
+        log.info("There are %s paths in the local path cache." % len(pc_data))
+        
+        log.info("")
+        log.info("Starting migration.")
+        
+        sg_records = []
+
+        for sql_record in pc_data:
+            log.debug("Processing db record %s..." % str(sql_record))
+            
+            # resolve a local path from a root and a generic path
+            root_name = sql_record[4]
+            db_path = sql_record[5]
+            root_path = self._roots.get(root_name)
+            if not root_path:
+                # The root name doesn't match a recognized name, so skip this entry
+                log.debug("Skipping path '%s %s' which doesn't have a valid root." % (root_name, db_path))
+                continue
+            
+            local_os_path = self._dbpath_to_path(root_path, db_path)
+            
+            log.info("Processing '%s'..." % local_os_path)
+            
+            entity_type = sql_record[1]
+            entity_id = sql_record[2]
+            
+            # see if we have this in shotgun already
+            sg_dict_key = (local_os_path, entity_type, entity_id)
+            
+            # now check if we have any entry in the path cache already which has that path
+            if sg_dict_key in sg_existing_data:
+                log.info(" - Skipping as path is already in Shotgun.\n")
+                log.debug("Path '%s' (%s %s) is already in shotgun (id %s)" % (local_os_path, 
+                                                                               entity_type, 
+                                                                               entity_id, 
+                                                                               sg_existing_data[sg_dict_key]))
+                continue
+            
+            # ok so this path cache record needs to be pushed to shotgun. Check that 
+            # the linked entity actually exists in shotgun
+            log.debug("Looking up %s %s in Shotgun to verify it exists..." % (entity_type, entity_id))
+            sg_data = self._tk.shotgun.find_one(entity_type, [["id", "is", entity_id]])
```

Comment:
This could be very slow (the previous version wasn't doing this check I don't think?) - could the ids be collected and then found in one go?